File: packages/junoguard/src/feed.ts
/**
 * The live decision feed behind `juno watch`.
 *
 * Live mode derives events from `GET /v1/guard/status`, which is the only
 * polling surface the frozen contract gives us: a rise in `blocked_today` is a
 * block, a rise in `open_incidents` is an incident, a rise in
 * `spend_today_usd` is billable traffic. Offline mode replays a scripted run
 * so the feed still moves with no gateway.
 */

import type { Line, Style } from "./render.js";
import type { StatusResult } from "./types.js";

const LANE_W = 10;
const DECISION_W = 7;

export interface Event {
  decision: "allow" | "flag" | "block" | "incident";
  lane: string;
  detail: string;
}

const STYLES: Record<Event["decision"], [Style, string]> = {
  allow: ["allow", "allow"],
  flag: ["flag", "FLAG"],
  block: ["block", "BLOCK"],
  incident: ["block", "!"],
};

export function renderEvent(event: Event, clock: string): Line {
  const [style, label] = STYLES[event.decision] ?? (["dim", event.decision] as [Style, string]);
  const detailStyle: Style = event.decision === "allow" ? "dim" : style;
  return [
    [`${clock}  `, "dim"],
    [label.padEnd(DECISION_W), style],
    [event.lane.padEnd(LANE_W), "key"],
    [event.detail, detailStyle],
  ];
}

export function clockNow(): string {
  const now = new Date();
  return [now.getHours(), now.getMinutes(), now.getSeconds()]
    .map((part) => String(part).padStart(2, "0"))
    .join(":");
}

// --------------------------------------------------------------------------
// live mode — status deltas
// --------------------------------------------------------------------------

export function diffStatus(previous: StatusResult, current: StatusResult): Event[] {
  const events: Event[] = [];

  if (previous.status !== current.status) {
    events.push({
      decision: current.status === "active" ? "allow" : "incident",
      lane: "project",
      detail: `project is now ${current.status}`,
    });
  }

  const blockedDelta = (current.blocked_today ?? 0) - (previous.blocked_today ?? 0);
  for (let i = 0; i < blockedDelta; i += 1) {
    events.push({ decision: "block", lane: "guard", detail: "an action was blocked by policy" });
  }

  const incidentDelta = (current.open_incidents ?? 0) - (previous.open_incidents ?? 0);
  for (let i = 0; i < incidentDelta; i += 1) {
    events.push({ decision: "incident", lane: "incident", detail: "new incident opened" });
  }

  const spendDelta = (current.spend_today_usd ?? 0) - (previous.spend_today_usd ?? 0);
  if (spendDelta > 0) {
    const budget = current.daily_budget_usd ?? 0;
    const remaining = current.remaining_usd ?? Math.max(0, budget - (current.spend_today_usd ?? 0));
    events.push({
      decision: "allow",
      lane: "llm",
      detail: `+$${spendDelta.toFixed(6)}  ·  $${remaining.toFixed(4)} left today`,
    });
  }

  const rpm = current.requests_last_min;
  const cap = current.max_requests_per_min;
  const wasOver =
    previous.requests_last_min !== undefined &&
    previous.max_requests_per_min !== undefined &&
    previous.requests_last_min > previous.max_requests_per_min;
  if (rpm !== undefined && cap && rpm > cap && !wasOver) {
    events.push({
      decision: "flag",
      lane: "burst",
      detail: `${rpm} req/min against a ${cap}/min limit`,
    });
  }

  return events;
}

// --------------------------------------------------------------------------
// offline mode — scripted run
// --------------------------------------------------------------------------

export const SCRIPT: Event[] = [
  { decision: "allow", lane: "install", detail: "express (npm) · clean" },
  { decision: "allow", lane: "llm", detail: "gpt-4o · 1,204 in / 287 out · $0.000431" },
  { decision: "allow", lane: "install", detail: "zod (npm) · clean" },
  {
    decision: "flag",
    lane: "install",
    detail: "@ossprey/suspicious-package (npm) · no published provenance",
  },
  { decision: "allow", lane: "llm", detail: "gpt-4o · 842 in / 190 out · $0.000312" },
  {
    decision: "block",
    lane: "install",
    detail: "@ossprey/test-package (npm) · malicious: obfuscated postinstall",
  },
  {
    decision: "incident",
    lane: "incident",
    detail: "critical · attempted credential exfiltration on install",
  },
  { decision: "allow", lane: "llm", detail: "gpt-4o · 611 in / 140 out · $0.000253" },
  { decision: "flag", lane: "burst", detail: "63 req/min against a 60/min limit" },
];

File: cli/juno_cli/feed.py
"""The live decision feed behind `juno watch`.

Live mode derives events from `GET /v1/guard/status`, which is the only polling
surface the frozen contract gives us: a rise in `blocked_today` is a block, a
rise in `open_incidents` is an incident, a rise in `spend_today_usd` is billable
traffic. Offline mode replays a scripted run so the feed still moves on stage.
"""

from __future__ import annotations

import time
from dataclasses import dataclass

from .render import S_ALLOW, S_BLOCK, S_DIM, S_FLAG, S_KEY, Line

_LANE_W = 10
_DECISION_W = 7


@dataclass
class Event:
    decision: str  # allow | flag | block | incident
    lane: str  # install | llm | project
    detail: str


_STYLES = {
    "allow": (S_ALLOW, "allow"),
    "flag": (S_FLAG, "FLAG"),
    "block": (S_BLOCK, "BLOCK"),
    "incident": (S_BLOCK, "!"),
}


def render_event(event: Event, clock: str) -> Line:
    style, label = _STYLES.get(event.decision, (S_DIM, event.decision))
    detail_style = S_DIM if event.decision == "allow" else style
    return [
        (f"{clock}  ", S_DIM),
        (label.ljust(_DECISION_W), style),
        (event.lane.ljust(_LANE_W), S_KEY),
        (event.detail, detail_style),
    ]


def clock_now() -> str:
    return time.strftime("%H:%M:%S")


# --------------------------------------------------------------------------
# live mode — status deltas
# --------------------------------------------------------------------------


def diff_status(previous: dict, current: dict) -> list[Event]:
    events: list[Event] = []

    if previous.get("status") != current.get("status"):
        state = current.get("status")
        events.append(
            Event(
                "incident" if state != "active" else "allow",
                "project",
                f"project is now {state}",
            )
        )

    blocked_delta = (current.get("blocked_today") or 0) - (previous.get("blocked_today") or 0)
    for _ in range(max(0, blocked_delta)):
        events.append(Event("block", "guard", "an action was blocked by policy"))

    incident_delta = (current.get("open_incidents") or 0) - (previous.get("open_incidents") or 0)
    for _ in range(max(0, incident_delta)):
        events.append(Event("incident", "incident", "new incident opened"))

    spend_delta = (current.get("spend_today_usd") or 0.0) - (previous.get("spend_today_usd") or 0.0)
    if spend_delta > 0:
        budget = current.get("daily_budget_usd") or 0.0
        remaining = current.get("remaining_usd")
        if remaining is None:
            remaining = max(0.0, budget - (current.get("spend_today_usd") or 0.0))
        events.append(
            Event("allow", "llm", f"+${spend_delta:.6f}  ·  ${remaining:.4f} left today")
        )

    rpm = current.get("requests_last_min")
    cap = current.get("max_requests_per_min")
    was_over = (previous.get("requests_last_min") or 0) > (previous.get("max_requests_per_min") or 10**9)
    if rpm is not None and cap and rpm > cap and not was_over:
        events.append(Event("flag", "burst", f"{rpm} req/min against a {cap}/min limit"))

    return events


# --------------------------------------------------------------------------
# offline mode — scripted run
# --------------------------------------------------------------------------

SCRIPT: list[Event] = [
    Event("allow", "install", "express (npm) · clean"),
    Event("allow", "llm", "gpt-4o · 1,204 in / 287 out · $0.000431"),
    Event("allow", "install", "zod (npm) · clean"),
    Event("flag", "install", "@ossprey/suspicious-package (npm) · no published provenance"),
    Event("allow", "llm", "gpt-4o · 842 in / 190 out · $0.000312"),
    Event("block", "install", "@ossprey/test-package (npm) · malicious: obfuscated postinstall"),
    Event("incident", "incident", "critical · attempted credential exfiltration on install"),
    Event("allow", "llm", "gpt-4o · 611 in / 140 out · $0.000253"),
    Event("flag", "burst", "63 req/min against a 60/min limit"),
]
